python/tritonserver/_c/tritonserver_pybind.cc

```
@@ -896,107 +871,120 @@ class PyInferenceRequest
+    //       static initialized?
+    static std::once_flag allocator_init_once;
+    std::call_once(allocator_init_once, [this]() {
+      struct TRITONSERVER_ResponseAllocator* allocator_raw = nullptr;
+      ThrowIfError(TRITONSERVER_ResponseAllocatorNew(
+          &allocator_raw, ResponseAllocatorAllocFn, ResponseAllocatorReleaseFn,
+          nullptr /* start_fn */));
+      allocator_.reset(allocator_raw);
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+          allocator_.get(), ResponseAllocatorQueryFn));
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+          allocator_.get(), ResponseAllocatorBufferAttributesFn));
+    });
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
   {
-    request_callback_resource_.reset(
-        new CallbackResource(release, user_object));
-    ThrowIfError(TRITONSERVER_InferenceRequestSetReleaseCallback(
-        triton_object_, PyTritonRequestReleaseCallback,
-        request_callback_resource_.get()));
+    // TODO: Support GPU memory.
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
   }
-
-  static void PyTritonRequestReleaseCallback(
-      struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
-      void* userp)
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    py::gil_scoped_acquire gil;
-    auto cr = reinterpret_cast<CallbackResource*>(userp);
-    cr->release_fn(cr->request, flags, cr->user_object);
-    delete cr;
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
```

do we want to throw if error or return the error?

---

sure, we can return the error. [Return if error when setting buffer attributes](https://github.com/triton-inference-server/core/pull/426/commits/c534d3a8a9e2b594b5e584f7c37b151a381fdcbf)